### DISCUSSION TITLE
Default EditorTextField control in a separate plugin to Normal mode instead of Insert mode

### DISCUSSION THREAD

**Discussion by AlexGirardDev** [JetBrains/ideavim#639] [Q&A] [answered]
Hey,

So I develop a plugin named [HarpoonIJ](https://github.com/AlexGirardDev/HarpoonIJ), while it is not directly related to IdeaVim alot of its users use them together.

The interaction I am having an issue with is that currently I open a popup dialog with a EditorTextField in it.
This editor defaults to being in Insert mode and I would like it to default to Normal mode. 

Is there a way for me to configure this behavior for only this text field?

Thanks.

**Comment by AlexPl292** [comment]
Hi there. Is the vim mode works in this dialog at all? I believe it was some optimization for the dialogs to start with the insert mode.

**Reply by AlexGirardDev** [reply]
Yes, vim mode does work in the dialog its just that you are defaulted into insert mode and i would like it to use normal mode.

I think there is a setting to set the default mode of all dialog editors which i don't want to use because in most situations i would want to start in insert mode for most things in the editor.

**Reply by AlexPl292** [reply]
I'm sure that this will be complicated (if even possible) to make a per-dialog solution. We can either enable normal mode everywhere or use insert mode everywhere in dialogs. Could you please describe if you remember any dialogs where the insert mode is more comfortable for you?

**Reply by HolbyFPV** [reply]
Any update on this?  Would ease a lot of friction if dialog windows started in normal mode.  
@AlexGirardDev @AlexPl292

**Reply by AlexGirardDev** [reply]
Honsestly I can't really. I am having a hard time thinking of a dialog that i use that popsup that actualy just allows free text entry.

i know in some tool windows i use starting in insert mode does make mores sense though.
like the AI assistant chat window
![image](https://github.com/JetBrains/ideavim/assets/37006998/9f0fff28-acde-453e-b028-d9a1ff3e1c7d)

or entering a commit message / a PR comment 

![image](https://github.com/JetBrains/ideavim/assets/37006998/09fb67c6-9f33-468e-8400-3f2ff295b185)

i wonder if there is some kind of workaround i could use. Is there some VimPlugin api i can use to force the mode to normal once the dialog has been opened?

**Reply by AlexPl292** [reply]
You may try to use `com.maddyhome.idea.vim.api.VimEditor#exitInsertMode` function.

**Reply by HolbyFPV** [reply]
Trying to call it like this inside `createCenterPanel()` within [HarpoonDialog](https://github.com/AlexGirardDev/HarpoonIJ/blob/master/src/main/java/ca/alexgirard/harpoonIJ/HarpoonDialog.java) is not working.  
```java
((VimEditor) (editorTextField.getEditor())).exitInsertMode(this, new OperatorArguments(false, 0, VimStateMachine.Mode.COMMAND, VimStateMachine.SubMode.NONE));
```
I think I'm on the wrong track.  Struggling to understand `executionContext` and if this cast is the right approach.

**Comment by HolbyFPV** [comment]
This looks a little more correct:  
```java
            OperatorArguments args = new OperatorArguments(false, 0, VimStateMachine.Mode.VISUAL, VimStateMachine.SubMode.NONE);
            ModeHelper.exitInsertMode(editorTextField.getEditor(), editor.getDataContext(), args);
```
But I can't get around a ClassNotFound exception when calling anything from the vim plugin API.  The IDE is finding the file just fine, but at runtime it's not found.  @AlexGirardDev is stuck on this exception as well.    

@AlexPl292 If we are unable to work around this in this way, would you be open to adding a setting to the vim config to set either insert mode everywhere or normal mode everywhere like you suggested in your first comment?

**Reply by AlexPl292** [reply, marked as answer]
Answered here: https://github.com/AlexGirardDev/HarpoonIJ/issues/2#issuecomment-1703068450

**Comment by AlexGirardDev** [comment]
Resloved, thank you for the help
